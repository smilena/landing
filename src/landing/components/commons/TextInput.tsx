import { useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import styled from "styled-components";
import { setUserData, UserData } from '../../../redux-state/UserDataSlice';
import { setContinueToNavigation } from "../../../redux-state/NavigationSlice";
import { useEffect } from "react";
import { RootState } from '../../../store';

export type TextInputProps = {
  id: string;
  label: string;
  maxlength?: number;
}

const Components = {
  Wrapper: styled.div`
    display: flex;
    flex-direction: column;
  `,
  Label: styled.label`
    color: #504437;
    font-size: 18px;
  `,
  Input: styled.input`
    color: #504437;
    font-size: 18px;
    height: 30px;
    border: 1px solid #504437;
    border-radius: 3px;
  `,
}

export const TextInput = ({id, label, maxlength}: TextInputProps) => {

  const dispatch = useDispatch();
  const userData: UserData = useSelector((state: RootState) => state.userData);
  const [value, setValue] = useState(userData[id  as keyof UserData ] as string);

  useEffect(() => {
    dispatch(setContinueToNavigation(value?.length > 0));
  }, [value]);

  const onChangeHandler = (e: React.FormEvent<EventTarget>) => {
    const { value } = e.target as HTMLInputElement;
    setValue(value);
    dispatch(setUserData({id, value}));
  }

  return (
    <Components.Wrapper>
      <Components.Label htmlFor={id}>{label}</Components.Label>
      <Components.Input
        data-testid="input-text"
        type="text" 
        id={id} 
        maxLength={maxlength} 
        value={value}
        onChange={onChangeHandler}
      />
    </Components.Wrapper>
  )
}
