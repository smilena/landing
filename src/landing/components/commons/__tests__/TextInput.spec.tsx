import { render, cleanup, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import createMockStore from 'redux-mock-store';
import { TextInput } from '../TextInput';

const mockStore = createMockStore([]);

afterEach(cleanup);

describe('Given TextInput component', () => {
  
  test('should render correctly with initial state', () => {
    const state = {
      userData: {
        name: 'Sandra',
      }
    };
    const store = mockStore(state);
    const { container } = render(
      <Provider store={store}>
        <TextInput id={'name'} label={'Nombres y apellidos'}/>
      </Provider>);
    expect(container).toMatchSnapshot();
    expect(screen.getByTestId('input-text').getAttribute('value')).toEqual('Sandra');
    expect(screen.getByText('Nombres y apellidos')).toBeInTheDocument();
  });
});
