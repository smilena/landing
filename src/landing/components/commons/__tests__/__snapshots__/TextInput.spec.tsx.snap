// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given TextInput component should render correctly with initial state 1`] = `
<div>
  <div
    class="sc-bczRLJ eaUlQc"
  >
    <label
      class="sc-gsnTZi bfxakb"
      for="name"
    >
      Nombres y apellidos
    </label>
    <input
      class="sc-dkzDqf dWwYGf"
      data-testid="input-text"
      id="name"
      type="text"
      value="Sandra"
    />
  </div>
</div>
`;
